### ISSUE TITLE
Github-based CI is extremely unreliable

### ISSUE THREAD

**Issue by bearbin** [cuberite/cuberite#4553]
Builds are failing randomly for various reasons:

 * AppVeyor: Network failure, random crashes during compilation
 * Travis: random freezes, failure to report success
 * CircleCI: cancelled for no apparent reason

Someone needs to see about resolving these. Also, as well as being slow in the best-case, appVeyor and Travis are accumulating large backlogs approaching days long when things are being actively developed, we need to see if there's a way to reduce these.

**Comment by madmaxoft** [comment]
Appveyor should at least cancel builds for PRs when they receive a new commit. No idea if Travis can be set up similarly.

**Comment by peterbell10** [comment]
>Travis: random freezes, failure to report success

Travis has had a few issues site wide over the last while.
https://www.traviscistatus.com/
Or do you mean cuberite specific build issues?

>Appveyor should at least cancel builds for PRs when they receive a new commit. No idea if Travis can be set up similarly.

Travis says it has auto-cancellation enabled
https://travis-ci.org/github/cuberite/cuberite/settings

**Comment by bearbin** [comment]
> Travis has had a few issues site wide over the last while.
> https://www.traviscistatus.com/
> Or do you mean cuberite specific build issues?

No, it appears the issues we've been having do correspond to the ones reported on their status page. Hopefully it should be better from now onwards.

The auto-cancellation does seem to generate another issue, when builds are cancelled they become nice red xs. But there's not really any way around that that I can see.

**Comment by bearbin** [comment]
I looked into AppVeyor, which seems to be the most-backlogged CI at the moment. It seems that it's easily possible to add a self-hosted Windows computer to do more builds; if any of the developers have a spare windows PC setting that up would be a good idea to get the backlog under control.

**Comment by bearbin** [comment]
I've disabled redundant build cancellation for CircleCI as it is not a bottleneck and we have more than enough credits not to need it. Hopefully that should stop the non-redundant builds on master from being cancelled.

**Comment by bearbin** [comment]
The issue with travis seems to be that we just have way too many individual jobs in each build and as a result the builds take too long (1 hour). Ideally we should reduce the build time to <10 minutes; whether by making the individual jobs faster or reducing the number of redundant ones.

To some extent the too-long-builds problem is also the case with Appveyor; if we could make those builds faster as well that would be optimal.

**Comment by peterbell10** [comment]
For appveyor I've enabled "rolling build" aka auto cancellation. I also saw "Do not build feature branches with open Pull Requests" which should stop PRs from branches on the main repo being built twice.

For travis, the slowest jobs are macOS because they don't use `ccache`.

**Comment by madmaxoft** [comment]
As for the Appveyor self-hosted builds, we could try the Azure free-for-12-months offer - that comes with a B1S Windows VM, which seems compatible with Appveyor. Now who's definitely NOT using their 12 month free offer with Azure, so that they can donate it to building Cuberite? :)

**Comment by madmaxoft** [comment]
And perhaps we should reconsider whether we want both Release and Debug builds on the CI.

**Comment by bearbin** [comment]
CircleCI and travis now appear to have returned to normal; although travis builds are still really too slow.

I have already used my free credit on Azure so that wouldn't be a solution for AppVeyor.

**Comment by bibo38** [comment]
> As for the Appveyor self-hosted builds, we could try the Azure free-for-12-months offer - that comes with a B1S Windows VM, which seems compatible with Appveyor. Now who's definitely NOT using their 12 month free offer with Azure, so that they can donate it to building Cuberite? :)

Have you considered [GitHub Actions](https://github.com/features/actions)? It supports Windows/Linux and macOS Runners. Btw. the issue title looks a bit misleading, as I first thought it was about GitHub Actions being unreliable.

**Comment by peterbell10** [comment]
>The issue with travis seems to be that we just have way too many individual jobs in each build and as a result the builds take too long (1 hour). Ideally we should reduce the build time to <10 minutes

I think saying 1 hour is a bit misleading. Travis runs most jobs in parallel so it was more like 30 minutes from commit to completion. With `ccache` on macOS as well now, most builds return in 10-15 minutes.

**Comment by bearbin** [comment]
I didn't even know GitHub Actions existed/could be used for this, assuming they give us more than one parallel executor, that seems like a good service to experiment with.

> I think saying 1 hour is a bit misleading. Travis runs most jobs in parallel so it was more like 30 minutes from commit to completion. With ccache on macOS as well now, most builds return in 10-15 minutes.

Yes, the time now is acceptable; but the builds don't always run with 8x parallelism (especially when load is high) so the reported time can actually be the real time.

**Comment by bibo38** [comment]
> I didn't even know GitHub Actions existed/could be used for this, assuming they give us more than one parallel executor, that seems like a good service to experiment with.

GitHub Actions allows up to 20 concurrent jobs with the exception that macOS jobs are limited to 5 concurrent jobs. Source: https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits

**Comment by mathiascode** [comment]
I actually experimented a bit with Github Actions, when looking into whether or not we could serve MacOS downloads from it. Didn't really work out for that purpose, since downloads have been limited to registered GitHub users, and don't have a dynamic URL, but here's a sample setup: https://github.com/mathiascode/cuberite/blob/f2760105871ec6f0d997d34252532b4c300aa68c/.github/workflows/macos.yml

**Comment by bibo38** [comment]
> Didn't really work out for that purpose, since downloads have been limited to registered GitHub users, and don't have a dynamic URL

Wouldn't it be possible to push the builds to the Github Releases (and maybe just override the old ones to prevent cluttering the Releases)? Never used the Actions but a quick search found this: https://github.com/marketplace/actions/github-release-create-update-and-upload-assets

**Comment by peterbell10** [comment]
I don't think publishing a new release for every commit to master is what github had in mind.

**Comment by bibo38** [comment]
Just tested Github Actions on my fork and they were really easy to set up. Just configure them with the yml file and they automatically integrate into Pull Requests and normal commits.

Here is the default yml file with about 3 lines changed to build cuberite: https://github.com/bibo38/Cuberite/blob/master/.github/workflows/main.yml

And it just works for Pull-Request: https://github.com/bibo38/Cuberite/pull/1
(the CI output can be found under the `Checks` tab in the pull request)

**Comment by madmaxoft** [comment]
I've pinged Appveyor with the problem, asking if there's anything we can do. They just replied that they've performed some maintenance and the issue should go away now.
https://help.appveyor.com/discussions/problems/26686-builds-randomly-failing-with-unexpected-logger-failure

**Comment by bearbin** [comment]
It seems the issue with Appveyor was not resolved by whatever 'maintenance' they have performed, as evidenced by this recent failure: https://ci.appveyor.com/project/Cuberite/cuberite/builds/31831658/job/acrbk8q4p8iriwjd

**Comment by tigerw** [comment]
We could consider reducing the number of configurations in the build matrix.

**Comment by tigerw** [comment]
Seems to be ok now.

**Comment by bearbin** [comment]
Currently seems to be working fine, but there hasn't been a high-intensity working period recently. I'll say things are good enough for now and close this.
